dev-support/design-docs/HBASE-23347-pluggable-authentication.md

```diff
@@ -0,0 +1,174 @@
+3. Client selection logic to choose a provider (from many that may be available)
+
+A provider's client and server side logic are considered to be one-to-one. A `Foo` client-side provider
+should never be used to authenticate against a `Bar` server-side provider.
+
+We do expect that both clients and servers will have access to multiple providers. A server may
+be capable of authenticating via methods which a client is unaware of. A client may attempt to authenticate
+against a server which the server does not know how to process. In both cases, the RPC
+should fail when a client and server do not have matching providers. The server identifies
+client authentication mechanisms via a `byte authCode` (which is already sent today with HBase RPCs).
+
+A client may also have multiple providers available for it to use in authenticating against
+HBase. The client must have some logic to select which provider to use. Because we are
+allowing custom providers, we must also allow a custom selection logic such that the
+correct provider can be chosen. This is a formalization of the logic already present
+in `org.apache.hadoop.hbase.security.token.AuthenticationTokenSelector`.
+
+To enable the above, we have some new interfaces to support the user extensibility:
+
+1. `interface SaslAuthenticationProvider`
+2. `interface SaslClientAuthenticationProvider extends SaslAuthenticationProvider`
+3. `interface SaslServerAuthenticationProvider extends SaslAuthenticationProvider`
+4. `interface AuthenticationProviderSelector`
+
+The `SaslAuthenticationProvider` shares logic which is common to the client and the
+server (though, this is up to the developer to guarantee this). The client and server
+interfaces each have logic specific to the HBase RPC client and HBase RPC server
+codebase, as their name implies. As described above, an implementation
+of one `SaslClientAuthenticationProvider` must match exactly one implementation of
+`SaslServerAuthenticationProvider`. Each Authentication Provider implementation is
+a singleton and is intended to be shared across all RPCs. A provider selector is
+chosen per client based on that client's configuration.
+
+A client authentication provider is uniquely identified among other providers
+by the following characteristics:
+
+1. A name, e.g. "KERBEROS", "TOKEN"
+2. A byte (a value between 0 and 255)
+
+In addition to these attributes, a provider also must define the following attributes:
+
+3. The SASL mechanism being used.
+4. The Hadoop AuthenticationMethod, e.g. "TOKEN", "KERBEROS", "CERTIFICATE"
+5. The Token "kind", the name used to identify a TokenIdentifier, e.g. `HBASE_AUTH_TOKEN`
+
+It is allowed (even expected) that there may be multiple providers that use `TOKEN` authentication.
+
+### Factories
+
+To ease development with these unknown set of providers, there are two classes which
+find, instantiate, and cache the provider singletons.
+
+1. Client side: `class SaslClientAuthenticationProviders`
```

Comment:
Great question! I put SASL in there to outwardly "align" any of these methods with a SASL mechanism. Meaning, anything you can express as a SASL mechanism should be easy to embed in here. Theoretically, we could expand this out further, but it would require major rethinking of our RPC code.

As someone a little farther away from the change: do you think dropping SASL is good? It could just be documentation on the interface itself explaining that SASL is what's used.